linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,117 @@ pub mod add_origin {
+                }
+            }
         }
     }
 
-    // === impl Init ===
+    // === impl MakeFuture ===
 
-    impl<M> Future for Init<M>
+    impl<S> Future for MakeFuture<S>
     where
-        M: tower::Service<client::Target>,
+        S: tower::Service<NameAddr>,
+        S::Response: Stream<Item = Result<Update<SocketAddr>, Error>> + Send + Sync + 'static,
+        S::Error: Into<Error> + Send + Sync,
+        S::Future: Send + Sync,
     {
-        type Output = Result<M::Response, Error<M::Error>>;
+        type Output = Result<UpdatesStream, S::Error>;
 
         fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-            let mut this = self.project();
-            loop {
-                match this.state.as_mut().project() {
-                    StateProj::Resolve(fut, stack) => {
-                        let ip = ready!(fut.poll(cx).map_err(Error::Dns))?;
-                        let (svc, config) = stack.take().unwrap();
-                        let addr = SocketAddr::from((ip, config.addr.port()));
-                        this.state
-                            .as_mut()
-                            .set(State::NotReady(svc, Some((addr, config))));
-                    }
-                    StateProj::NotReady(svc, cfg) => {
-                        ready!(svc.poll_ready(cx).map_err(Error::Inner))?;
-                        let (addr, config) = cfg.take().unwrap();
-                        let state = State::make_inner(addr, &config, svc);
-                        this.state.as_mut().set(state);
-                    }
-                    StateProj::Inner(fut) => return fut.poll(cx).map_err(Error::Inner),
-                };
-            }
-        }
-    }
-
-    impl<M> State<M>
-    where
-        M: tower::Service<client::Target>,
-    {
-        fn make_inner(addr: SocketAddr, dst: &ControlAddr, mk_svc: &mut M) -> Self {
-            let target = client::Target {
-                addr,
-                server_name: dst.identity.clone(),
-            };
-
-            State::Inner(mk_svc.call(target))
-        }
-    }
-
-    // === impl Error ===
-
-    impl<I: fmt::Display> fmt::Display for Error<I> {
-        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-            match self {
-                Error::Dns(dns::Error::NoAddressesFound) => write!(f, "no addresses found"),
-                Error::Dns(e) => fmt::Display::fmt(&e, f),
-                Error::Inner(ref e) => fmt::Display::fmt(&e, f),
-            }
+            let this = self.project();
+            let identity = this.identity.clone();
+            let stream = futures::ready!(this.inner.poll(cx))?;
+            Poll::Ready(Ok(Box::pin(stream.map_ok(move |update| {
+                match update {
+                    Update::Add(adds) => Update::Add(
+                        adds.clone()
+                            .into_iter()
+                            .map(|(sa, _)| (sa, Target::new(sa, identity.clone())))
```

Comment:
Curious if we can avoid having to have this code by using a resolve::MapEndpoint instead